### DISCUSSION TITLE
Dominando Injeção de Dependência -  Utilizando Serviços

### DISCUSSION THREAD

**Discussion by JoabeRochadel** [balta-io/forum#684] [Tira Dúvidas] [answered]
Boa tarde equipe Balta,

Sempre que estou implementando me surge 2 dúvidas cruciais, 1º Regras de negócio, onde devo colocar elas? É sempre considerado um Services? 2º Regras de negócio complexa, imagine uma regra de negócio para algum ramo onde tenha uma convenção/orgão fiscalizador e tenho uma regra de negócio complexa e grande. Devo quebrar em cada regra um metodo específico ou fazer ela quase de forma procedural?

**Comment by JoabeRochadel** [comment]
Pra exemplificar o que eu estou falando, podemos usar a lógica do DeliveryFeeServices
Imagine que faço a entrega para X países e cada país tem sua lógica de entregas, devo implementar todas nessa classe ou quebrar em classes menores gerando diversas classes pra esse papel.

**Reply by andrebaltieri** [reply, marked as answer]
Bom dia, @JoabeRochadel, como vai?

1º Regras de negócio, onde devo colocar elas? 
- Depende do modelo que está desenvolvendo, se for seguir o DDD, elas ficam no domínio (Outras abordagens como arquitetura limpa também vão te dizer isto... é a recomendação mais usada).

É sempre considerado um Services? 
- Não! Existem diferentes tipos de regras... veja o curso Refatorando para Testes de Unidade, lá testo até queries... ou seja, temos regras de leituras também.

2º Regras de negócio complexa, imagine uma regra de negócio para algum ramo onde tenha uma convenção/orgão fiscalizador e tenho uma regra de negócio complexa e grande. Devo quebrar em cada regra um metodo específico ou fazer ela quase de forma procedural?
- Sempre que puder especializar é melhor... menor, mais fácil de ler, mais simples de entender, mais fácil de testar 🚀

Veja se isto te ajuda!
